docs/blueprints/002-extension-registry.md

```diff
@@ -0,0 +1,112 @@
+workflow steps.
+
+## Current state
+
+There is currently no dedicated means of dynamically storing information on the FuseML server describing where and how to access
+3rd party tools and services. To compensate for this limitation, information such as 3rd party tool endpoint URLs and credentials
+is instead provided using other extensibility mechanisms:
+* in workflow definitions (as is currently the case with MLFlow)
+* hardcoded inside the container images implemeting workflow steps
+* hardcoded inside codesets
+* even hardcoded directly into the FuseML server itself
+
+This has significant negative impact on flexibiity and reusability, on user experience in general.
+
+For 3rd party tools implemented as k8s controllers, such as KNative, this is less of a problem, because pods running as
+pipeline steps are by default given access to create and manage k8s resources in the workloads namespace through a custom
+service account. However, this will be more difficult to achieve when adding multi-cluster capabilities (see https://github.com/fuseml/fuseml/issues/146).
+
+## Feature Requirements
+
+The extension registry feature implements the following mandatory requirements:
+
+1. as an Ops engineer (MLOps/DevOps/ITOps), I need a way to configure my FuseML instance with the parameters required to dynamically integrate with 3rd party AI/ML tools (e.g. URLs, endpoints, credentials, other tool specific configuration attributes). For non-production environments, this requirement can also come from ML engineers or even Data Scientists that are looking to quickly set up FuseML for experimentation purposes.
+
+Examples:
+* DS: I have an MLFlow tracking service already set up by me that I use for my ML experiments and I want to reuse it for FuseML automated workflows. I will configure FuseML with the information it needs to access the MLFlow tracking service (the tracker service URL and the hostname, username and keys for the storage backend). The information is stored in a central registry where FuseML workflows can access it.
+* DS: I'm using my Google cloud storage account to store datasets or ML models that I use in my local experiments. I want my FuseML automated workflows to upload/download artifacts using that same storage account, but I don't want to expose my credentials in the workflow definition or in the code I'm pushing to FuseML. I'll store those credentials in the FuseML extension registry and access them from my FuseML workflow steps.
+* Ops: I have an S3 storage service set up for my organization and I want to use that as a storage backend for FuseML artifacts (e.g. models, datasets).
+I will manage buckets, accounts and credentials and add them as extension records in the FuseML extension registry. ML engineers and DSs can then write FuseML workflows that have access to the S3 storage service without having to deal with these operational details.
```

Comment:
Maybe you should add an indentation here? I got confused when switching from a dotted list to a numbered one that is also under examples, however it is a continuation of the requirements.

Reply:
Tougher than I thought, but done